File: scripts/lib/git-base.mjs
// Shared "resolve a base ref to diff against, on a possibly-shallow CI
// checkout" logic — originally written for constitution-check.mjs, factored
// out so build-relevant-changed.mjs doesn't duplicate it.
import { $ } from "bun";

async function tryResolve(root, candidate) {
  try {
    await $`git -C ${root} rev-parse --verify ${candidate}`.quiet();
    return true;
  } catch {
    return false;
  }
}

async function hasMergeBase(root, candidate) {
  try {
    await $`git -C ${root} merge-base ${candidate} HEAD`.quiet();
    return true;
  } catch {
    return false;
  }
}

// Resolves `explicitBase` if given, else tries local `origin/main`/`main`,
// else fetches `origin/main` explicitly (a PR-triggered CI checkout is
// typically shallow and only fetches the ref needed for the merge commit —
// `origin/main` may genuinely not be a resolvable local ref yet, not because
// there's no base to diff against). Once a base is found, ensures a
// merge-base with HEAD exists (deepening the checkout if not) so a
// three-dot diff — the one that actually answers "did *this branch* change
// X," not "does X currently differ from base's live tip" — is possible.
// Returns null if no base could be resolved at all.
export async function resolveBase(root, explicitBase) {
  let base = explicitBase ?? null;
  if (!base) {
    for (const candidate of ["origin/main", "main"]) {
      if (await tryResolve(root, candidate)) {
        base = candidate;
        break;
      }
    }
  }
  if (!base) {
    try {
      await $`git -C ${root} fetch origin main:refs/remotes/origin/main`.quiet();
    } catch {
      // network/remote unavailable — fall through to "no base" below
    }
    if (await tryResolve(root, "origin/main")) base = "origin/main";
  }
  if (!base) return null;

  if (!(await hasMergeBase(root, base))) {
    try {
      await $`git -C ${root} fetch --unshallow`.quiet();
    } catch {
      try {
        await $`git -C ${root} fetch --deepen=1000000`.quiet();
      } catch {
        // best effort — the caller's diff fails closed if this didn't help
      }
    }
  }
  return base;
}

// Three-dot diff (relative to the merge-base), matching resolveBase's own
// reasoning above.
export async function changedFiles(root, base) {
  const out = await $`git -C ${root} diff --name-only ${base}...HEAD`.text();
  return out.split("\n").filter(Boolean);
}

File: scripts/lib/git_base.py
"""Resolve a base ref to diff against, on a possibly-shallow CI checkout.

Shared by scripts/hygiene/constitution-check.py and
scripts/build-relevant-changed.py so neither duplicates this logic
(#166-style DRY).
"""

import subprocess


def _git(root, args):
    return subprocess.run(["git", "-C", str(root), *args], capture_output=True, text=True)


def _try_resolve(root, candidate):
    return _git(root, ["rev-parse", "--verify", candidate]).returncode == 0


def _has_merge_base(root, candidate):
    return _git(root, ["merge-base", candidate, "HEAD"]).returncode == 0


# Resolves `explicit_base` if given, else tries local `origin/main`/`main`,
# else fetches `origin/main` explicitly (a PR-triggered CI checkout is
# typically shallow and only fetches the ref needed for the merge commit --
# `origin/main` may genuinely not be a resolvable local ref yet, not because
# there's no base to diff against). Once a base is found, ensures a
# merge-base with HEAD exists (deepening the checkout if not) so a
# three-dot diff -- the one that actually answers "did *this branch* change
# X," not "does X currently differ from base's live tip" -- is possible.
# Returns None if no base could be resolved at all.
def resolveBase(root, explicit_base=None):
    base = explicit_base or None
    if not base:
        for candidate in ("origin/main", "main"):
            if _try_resolve(root, candidate):
                base = candidate
                break
    if not base:
        # network/remote unavailable -- fall through to "no base" below
        _git(root, ["fetch", "origin", "main:refs/remotes/origin/main"])
        if _try_resolve(root, "origin/main"):
            base = "origin/main"
    if not base:
        return None

    if not _has_merge_base(root, base):
        # best effort -- the caller's diff fails closed if this didn't help
        if _git(root, ["fetch", "--unshallow"]).returncode != 0:
            _git(root, ["fetch", "--deepen=1000000"])
    return base


# Three-dot diff (relative to the merge-base), matching resolveBase's own
# reasoning above.
def changedFiles(root, base):
    result = _git(root, ["diff", "--name-only", f"{base}...HEAD"])
    if result.returncode != 0:
        raise RuntimeError(
            f"git -C {root} diff --name-only {base}...HEAD failed (exit {result.returncode}): {result.stderr.strip()}"
        )
    return [line for line in result.stdout.split("\n") if line]
